feat: add note metadata

nventuro opened the pull request:
Follow up to https://github.com/AztecProtocol/aztec-packages/pull/11942. This improves the readability and management of code that reads and destroys notes by introducing the `NoteMetadata` struct, which handles the different combinations of note hash counter and nonce internally. This is a stepping stone towards #8589.

nventuro committed:
Add note metadata

nventuro committed:
Some docs

benesjan reviewed (approved):
Lovely. Just needs fixing some comments.

nventuro commented on noir-projects/aztec-nr/aztec/src/note/note_metadata.nr:
```diff
@@ -0,0 +1,187 @@
+}
+
+impl NoteMetadata {
+    /// Constructs a `NoteMetadata` object from optional note hash and nullifiers. Both a zero note hash counter and a
```
lol my brain just autocompleted here

nventuro committed:
Apply renamings

nventuro committed:
Fix test

nventuro committed:
Remove note hash counter

nventuro committed:
Fix from trait

nventuro committed:
Add comment

nventuro commented:
I adjusted some of the naming: we now correctly call same-tx notes 'pending' instead of 'transient', and clarified that the separation is not revertible vs non-revertible, but rather same phase or previous phase.

I also removed the note hash counter from the struct. This value was not being used anywhere, but was dangerous to have around since it was underconstrained. We only ever check if it is zero or non-zero to determine if the note is pending or not, but we never do validate its content (nor can we, since the read request does not include the counter, and we cannot inspect the new note hashes array). Therefore the only correct thing to do is to get rid of it.

nventuro committed:
Fix private exec test

benesjan commented on noir-projects/aztec-nr/aztec/src/note/utils.nr:
```diff
@@ -4,112 +4,86 @@ use crate::{
+            siloed_note_hash,
+        )
+    } else {
+        // Pending notes (both same phase and previous phase ones)  are read by their non-siloed hash (not even by
```
```suggestion
        // Pending notes (both same phase and previous phase ones) are read by their non-siloed hash (not even by
```
That extra space triggered me a bit.

benesjan reviewed (approved):
Looks good. Would just expand a bit in those 2 comments

nventuro commented on noir-projects/aztec-nr/aztec/src/note/note_interface.nr:
```diff
@@ -29,9 +29,16 @@ pub trait NullifiableNote {
+    /// requests) nor note hash (since it would be meaningless to tie this to a note that has been read).
+    ///
+    /// Second, it always compute the nullifier for a **settled** note, i.e. a note that has been created in a previous
+    /// transaction, which therefore has a nonce. This is typically fine, since this function will mostly be used in
```
After some reflection I think we just simplify this fn so that it also takes a note hash. I'll do that once both our PRs are in.

nventuro committed:
apply review comments

nventuro committed:
Merge branch 'master' into nv/note-metadata

iAmMichaelConnor commented:
Hey, we had unresolved arguments about this, from yesterday
